File: examples/ppocr/cpp/ppocr.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 *
 * PP-OCRv6 text detection + recognition example.
 */

#include <cstdio>
#include <iomanip>
#include <iostream>
#include <memory>
#include <string>
#include <variant>
#include <vector>

#include <opencv2/opencv.hpp>

#include "vision_service.h"

namespace {
void PrintUsage(const char* prog) {
    std::cout << "Usage: " << prog << " <config_yaml> [options]\n"
        << "  --model-path <path> Override det model_path in yaml\n"
        << "  --image <path>      Input image (overrides config test_image)\n"
        << "  --output <path>     Output image path (default: ppocr_result.jpg)\n"
        << "  --help              Show this help\n";
}
}  // namespace

int main(int argc, char* argv[]) {
    if (argc < 2) {
        PrintUsage(argv[0]);
        return 1;
    }

    std::string config_path = argv[1];
    std::string image_path;
    std::string output_path = "ppocr_result.jpg";
    std::string model_path_override;

    for (int i = 2; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--help") {
            PrintUsage(argv[0]);
            return 0;
        } else if (arg == "--image" && i + 1 < argc) {
            image_path = argv[++i];
        } else if (arg == "--output" && i + 1 < argc) {
            output_path = argv[++i];
        } else if (arg == "--model-path" && i + 1 < argc) {
            model_path_override = argv[++i];
        }
    }

    std::unique_ptr<VisionService> service = VisionService::Create(config_path, model_path_override, true);
    if (!service) {
        std::cerr << "Error: " << VisionService::LastCreateError() << std::endl;
        return 1;
    }

    if (image_path.empty()) {
        image_path = service->GetDefaultImage();
    }
    if (image_path.empty()) {
        std::cerr << "Error: No input image. Use --image or set test_image in config." << std::endl;
        return 1;
    }

    cv::Mat img = cv::imread(image_path);
    if (img.empty()) {
        std::cerr << "Error: Could not load image: " << image_path << std::endl;
        return 1;
    }

    VisionServiceResponse response;
    VisionServiceStatus ret = service->Infer(img, &response);
    if (ret != VISION_SERVICE_OK) {
        std::cerr << "Error: " << service->LastError() << std::endl;
        return 1;
    }

    if (!response.results.empty()) {
        std::cout << "Recognized " << response.results.size() << " text lines:" << std::endl;
        for (const auto& result : response.results) {
            const vision::Text* t = std::get_if<vision::Text>(&result);
            if (t == nullptr) {
                continue;
            }
            std::cout << "  \"" << t->text << "\"  score=" << std::fixed << std::setprecision(3)
                << t->score << "  quad=[";
            for (size_t i = 0; i < t->polygon.size(); ++i) {
                std::cout << "(" << static_cast<int>(t->polygon[i].x) << ","
                    << static_cast<int>(t->polygon[i].y) << ")";
            }
            std::cout << "]" << std::endl;
        }
        cv::Mat vis;
        if (service->Draw(img, response, &vis) != VISION_SERVICE_OK || vis.empty()) {
            std::cerr << "Draw failed: " << service->LastError() << "; saving original image"
                << std::endl;
            vis = img;
        }
        cv::imwrite(output_path, vis);
        std::cout << "Result image saved to: " << output_path << std::endl;
    } else {
        std::cout << "No text detected" << std::endl;
        cv::imwrite(output_path, img);
    }
    return 0;
}

File: examples/ppocr/python/ppocr.py
#!/usr/bin/env python3
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
PP-OCRv6 text detection + recognition example using spacemit_vision.

--config 指定 yaml → VisionServiceNative.create(config) 创建模型 → infer_image 出文本。
每个结果带 .text（识别文字）、.polygon（文本框四点）、.score（识别置信度）。
"""

import argparse
from pathlib import Path

import cv2

from spacemit_vision import VisionServiceNative, VisionServiceStatus


def resolve_path(path_value, project_root):
    p = Path(path_value).expanduser()
    return p if p.is_absolute() else (project_root / p).resolve()


def parse_args():
    parser = argparse.ArgumentParser(description="PP-OCRv6 OCR Example")
    parser.add_argument("--config", type=str, default=None,
                        help="Config yaml path (default: examples/ppocr/config/ppocr.yaml)")
    parser.add_argument("--model-path", type=str, default=None, help="Override det model_path in yaml")
    parser.add_argument("--image", type=str, help="Input image path (default: config test_image)")
    parser.add_argument("--output", type=str, default="ppocr_result.jpg", help="Output image path")
    return parser.parse_args()


def main():
    args = parse_args()
    try:
        default_config = Path(__file__).parent.parent / "config" / "ppocr.yaml"
        config_path = Path(args.config) if args.config else default_config
        project_root = Path(__file__).parent.parent.parent.parent
        model_name = config_path.stem

        print(f"创建 {model_name} OCR 模型...")
        svc = VisionServiceNative.create(
            str(config_path),
            model_path_override=args.model_path or "",
        )

        if args.image:
            image_path = resolve_path(args.image, project_root)
        else:
            default_image = svc.get_default_image()
            if not default_image:
                raise ValueError("未提供 --image，且 config 无 test_image")
            image_path = Path(default_image)

        print(f"加载图像: {image_path}")
        image = cv2.imread(str(image_path))
        if image is None:
            raise ValueError(f"无法加载图像: {image_path}")

        print("运行 OCR...")
        status, results = svc.infer_image(image)
        if status != VisionServiceStatus.OK:
            raise RuntimeError(svc.last_error())

        if results:
            print(f"识别到 {len(results)} 行文字:")
            for r in results:
                text = getattr(r, "text", "")
                poly = getattr(r, "polygon", [])
                pts = [(int(p.x), int(p.y)) for p in poly]
                print(f'  "{text}"  score={r.score:.3f}  quad={pts}')
            if svc.supports_draw():
                st, out = svc.draw(image)
                result_image = out if st == VisionServiceStatus.OK else image
            else:
                result_image = image
            cv2.imwrite(args.output, result_image)
            print(f"结果图像已保存到: {args.output}")
        else:
            print("未识别到文字")
            cv2.imwrite(args.output, image)
            print(f"原始图像已保存到: {args.output}")

    except Exception as e:
        print(f"错误: {e}")
        return 1

    return 0


if __name__ == "__main__":
    import sys
    sys.exit(main())
